Project files have been added.

rastepanyan opened the pull request.

rastepanyan committed:
Add project

rastepanyan committed:
Project files have been added.

andrey-dnh commented on docker-compose.yml:
```diff
@@ -0,0 +1,39 @@
+      ports:
+            - "3306:3306"
+    phpmyadmin:
+     image: phpmyadmin/phpmyadmin
```
подравни си остъпа да в като в db. Държа много на правилното форматиране :)

andrey-dnh commented on dump/01-products.sql:
```diff
@@ -0,0 +1,29 @@
+CREATE TABLE IF NOT EXISTS `products` (
+	`product_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
+	`title` VARCHAR(30)  NOT NULL DEFAULT '',
+	`price` DECIMAL(7,2)  NOT NULL DEFAULT '99999.99',
```
чудесно. Пари трябва да се представят с decimal, а не с float

andrey-dnh commented on dump/01-products.sql:
```diff
@@ -0,0 +1,29 @@
+	`description` MEDIUMTEXT NOT NULL,
+	`date_of_creation` DATETIME NOT NULL,
+	PRIMARY KEY (`product_id`))
+	ENGINE = InnoDB CHARSET=utf8 COLLATE utf8_swedish_ci;
```
виж дали можеш да използваш utf8mb4

andrey-dnh commented on dump/01-products.sql:
```diff
@@ -0,0 +1,29 @@
+				 
+INSERT INTO `products` (`title`, `price`, `description`, `date_of_creation`)
+VALUES	('Shoes', 119.99, 'An item of footwear intended to protect and comfort the human foot.', NOW()),
+		('Blouse', 42.75, 'A loose-fitting upper garment that was formerly worn by women and children.', NOW()),
```
вместо now() вкарай някакви различни дати. Може и да не са ръчни, просто NOW() - INTERVAL 1 DAY, NOW - INTERVAL 2 DAY и т.н.

andrey-dnh commented on dump/01-products.sql:
```diff
@@ -0,0 +1,29 @@
+USE mytask;
+
+CREATE TABLE IF NOT EXISTS `products` (
+	`product_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
```
Много добре! Челочисленият тип за сурогатния ключ е добре да е неотрицателно число, понеже MySQL никога няма да генерира отрицателни числа.

andrey-dnh commented on dump/02-users.sql:
```diff
@@ -0,0 +1,28 @@
+	`first_name` VARCHAR(20)  NOT NULL DEFAULT '',
+	`last_name` VARCHAR(30)  NOT NULL DEFAULT '',
+	`address` VARCHAR(255)  NOT NULL DEFAULT '',
+	`country_code` varchar(2) NOT NULL DEFAULT '',
```
Този стринг винаги ще съдържа два символа. Виж дали има по-подходящ тип.

andrey-dnh commented on dump/02-users.sql:
```diff
@@ -0,0 +1,28 @@
+	`user_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
+	`first_name` VARCHAR(20)  NOT NULL DEFAULT '',
+	`last_name` VARCHAR(30)  NOT NULL DEFAULT '',
+	`address` VARCHAR(255)  NOT NULL DEFAULT '',
```
В заданието "adress" е уловка. Нищо. Моля те добави отделно:
-  ПК - няма как да е число. Английските ПК са стрингове. Приеми, че дължината на английския ПК е най-дългия възможен ПК.
- град

andrey-dnh commented on dump/03-orders.sql:
```diff
@@ -0,0 +1,37 @@
+	`order_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
+	`order_date` DATETIME NOT NULL,
+	`user_id` INT UNSIGNED NOT NULL,
+	`amount` DECIMAL(7,2) NOT NULL DEFAULT '99999.99',
```
Цените са DECIMAL(7,2). Няма как сумата да е само 7,2 . Разшири малко типа.

andrey-dnh commented on dump/03-orders.sql:
```diff
@@ -0,0 +1,37 @@
+					 ('2018-07-20 18:00:12', 8, 12.89),
+					 ('2018-07-21 08:12:43', 9, 30.98),
+					 ('2018-07-21 10:36:28', 10, 206.67),
+					 ('2018-07-21 13:55:21', 11, 199.98),
```
разпръсни малко повече датите на поръчките, за да може заявката за последния месец да филтрира част от резултатите.

andrey-dnh commented on dump/04-orderLines.sql:
```diff
@@ -0,0 +1,37 @@
+CREATE TABLE IF NOT EXISTS `orderlines` (
+	`orderlines_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
+	`order_id` INT UNSIGNED NOT NULL,
+	`price_each` DECIMAL(7,2) NOT NULL DEFAULT '99999.99',
```
Бих го нарекал item_price

andrey-dnh replied:
Един допълнителен индекс order_id, product_id ще помогне на определени заявки

andrey-dnh commented on dump/04-orderLines.sql:
```diff
@@ -0,0 +1,37 @@
+	`order_id` INT UNSIGNED NOT NULL,
+	`price_each` DECIMAL(7,2) NOT NULL DEFAULT '99999.99',
+	`product_id` INT UNSIGNED NOT NULL,
+	`quantity` INT UNSIGNED NOT NULL,
```
добре

andrey-dnh commented on dump/05-printing_last_hour.sql:
```diff
@@ -0,0 +1,3 @@
+/* Printing the list of products, created during the last hour. */
+																						
+SELECT title FROM products WHERE date_of_creation > SUBDATE( NOW(), INTERVAL 1 HOUR);
```
правилно. Може и DATE_SUB() или направо с израз : NOW - INTERVAL 1 HOUR

andrey-dnh commented on dump/06-printing_code.sql:
```diff
@@ -0,0 +1,3 @@
+/* Printing the list of users sorted by country code.*/
+																						
+SELECT first_name, last_name, country_code FROM users ORDER BY country_code;
```
правилно

andrey-dnh commented on dump/07-printing_orders.sql:
```diff
@@ -0,0 +1,3 @@
+/* Printing a simple list of orders.*/
+
+SELECT order_id, user_id, order_date, product_id, price_each, amount FROM orderlines INNER JOIN orders using (order_id);
```
в тази заявка не ти трябва orderlines. Простият списък идва само от orders. Заданието е : 
"Printing a simple list of orders (no order lines)."

andrey-dnh commented on dump/04-orderLines.sql:
```diff
@@ -0,0 +1,37 @@
+CREATE TABLE IF NOT EXISTS `orderlines` (
+	`orderlines_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
```
хубаво е да добавиш и позиция на линията в поръчката.

andrey-dnh commented on dump/08-printing_orderlines.sql:
```diff
@@ -0,0 +1,3 @@
+/* 	Printing the list orderlines with combined order information and user information.*/
+
+SELECT order_id, price_each, product_id, quantity, order_date, user_id, amount, first_name, last_name, address, country_code FROM orderlines INNER JOIN orders using (order_id) INNER JOIN users using (user_id) ORDER BY order_id;
```
В SQL, ако не е зададена сортировка, не може да се гарантира че всеки път резултатът ще е същият. В случая, се изписват orderlines и се сортира по order_id. Като добавиш orderlines.position, я използвай също за сортирането.

andrey-dnh commented on dump/09-printing_number_of_orders.sql:
```diff
@@ -0,0 +1,3 @@
+/* 	Printing the list of users with number of orders they have placed.*/
+
+SELECT first_name, last_name, users.user_id, COUNT(*) AS number_of_orders FROM users, orders WHERE users.user_id = orders.user_id GROUP BY first_name, last_name, user_id ORDER BY users.user_id;
```
Бих сложил users.user_id отпред, вместо след имената.
Възможно е и използването на INNER JOIN с USING или ON клауза. По-добре показва как се навързват таблиците, а WHERE клаузата остава за филтрирането (selection).

andrey-dnh replied:
Не трябва да групираш по име и фамилия

andrey-dnh commented on dump/10-printing_spent_in_total.sql:
```diff
@@ -0,0 +1,3 @@
+/* Printing the list of users with a column of how much money they have spent in this shop. */
+
+SELECT SUM(orders.amount) total, users.first_name, users.last_name FROM orders, users WHERE orders.order_id = users.user_id GROUP BY first_name, last_name ORDER BY total;
```
Групирай по друга колона. В магазина може да има два пъти Борис Христов.

andrey-dnh commented on dump/11-printing_more_than_three_users.sql:
```diff
@@ -0,0 +1,3 @@
+/* Printing the list of countries that have more than 3 users, sorted by this number in descending order. */
+																						
+SELECT country_code, COUNT(*) AS mttu FROM users GROUP BY country_code HAVING mttu >= 3 ORDER BY mttu DESC;
```
more than three users = mttu бих го кръстил total_users

andrey-dnh commented on dump/12-printing_orders_per_country.sql:
```diff
@@ -0,0 +1,3 @@
+/* Printing the list of countries with the number of orders per country, sorted by this number in descending order. */
+																						
+SELECT country_code, COUNT(*) AS nopc FROM users, orders WHERE users.user_id = orders.user_id GROUP BY country_code ORDER BY nopc DESC;
```
Тук пак може да се използва INNER JOIN. Може, не е задължително.

andrey-dnh commented on dump/15-printing_most_ordered.sql:
```diff
@@ -0,0 +1,2 @@
+/* Printing the 10 most ordered products of all times. */
+SELECT product_id, SUM(quantity) AS most_ordered FROM orderlines GROUP BY product_id ORDER BY most_ordered DESC LIMIT 10;
```
правилно

andrey-dnh commented on www/index.php:
```diff
@@ -0,0 +1,34 @@
+<html>
```
още не сме стигнали до PHP, но добре :)
За по-голямо удобство може да направиш следното. Започни файла с PHP код, където правиш връзката, пускаш заявката, проверяваш за грешки и чак тогава затваряш PHP кода и започва HTML-a. В самия HTML пак отваряш PHP за да можеш да врътнеш цикъла за резултатите. Така ще имаш минимално количество PHP код в HTML-a. Реално погледнато тогава HTML-a може да се изнесе в отделен файл, т.е. да е шаблон/темплейт, несъдържащ бизнес логика.
Така кодът ще е на ниво PHP до 2005-та година, преди появяването на фреймуърци, но пак виждаш основните компоненти и разделянето на бизнес логика от презентация.

rastepanyan committed:
Fix and add finctionality

DOCKER-COMPOSE.YML

1. docker-compose.yml was re-formatted

01-PRODUCTS.SQL

1. The collation ‘utf8’ was changed to ‘utf8mb4’
2. Insert day intervals – NOT WORKING
3. Expand the type ‘DECIMAL(7,2)’ to ‘DECIMAL(10,2)’


02-USERS.SQL

1. The collation ‘utf8’ was changed to ‘utf8mb4’
2. Replace VARCHAR with CHAR in ‘country_code‘ column
3. Fix: Add ‘post_code’ column
4. Fix: Add ‘city’ column


03-ORDERS.SQL

1. The collation ‘utf8’ was changed to ‘utf8mb4’
2. Expand the type ‘DECIMAL(7,2)’ to ‘DECIMAL(10,2)’
3. Shuffle the order dates


04-ORDERLINES.SQL

1. Change the name of ‘price_each to ‘item_price’
2. Fix: Add ‘position’ column


05-PRINTING_LAST_HOUR.SQL

1. Add another solutions as a comment


07-PRINTING_ORDERS.SQL

1. Fix: Script show only orders now.


08-PRINTING_ORDERLINES.SQL

1. Fix: Add ‘position’ column.


09-PRINTING_NUMBER_OF_ORDERS.SQL

1. Fix: Printing ‘user_id’ on first column.
2. Fix: Group by ‘user_id’ now.
3. INNER JOIN is used as a solution.


10-PRINTING_SPENT_IN_TOTAL

1. Fix: Group by ‘user_id’ now.
11-PRINTING_MORE_THAN_THREE_USERS.SQL

1. Rename ‘mttu’ column to ‘total_users’.


12-PRINTING_ORDERS_PER_COUNTRY.SQL

1. INNER JOIN is used as a solution.

rastepanyan committed:
Delete 04-orderLines.sql

rastepanyan committed:
Add files via upload

rastepanyan commented:
Здравей, Андрей!


Вече добавих променените файлове в GitHub.

Изпълних всички препоръки, които ми даде. Засега не съм пипал само
'index.php' файла, но ще го променя и него днес или утре.

За съжаление в първия скрипт 01-products.sql не можах да задам автоматично
са променя датата и засега използвам отново NOW()). Изпробвах функциите:

SUBDATE(NOW(), INTERVAL 1 DAY);

DATE_SUB(NOW(), INTERVAL 1 DAY);

NOW() - INTERVAL 1 DAY

MySQL дава грешка ERROR 1064 (42000).



Хубава и лека вечер от мен!

Поздрави :)

On 26 July 2018 at 20:38, Andrey Hristov <notifications@github.com> wrote:

> *@andrey-dnh* requested changes on this pull request.
> ------------------------------
>
> In docker-compose.yml
> <https://github.com/rastepanyan/training/pull/2#discussion_r205530497>:
>
> > +      image: mysql:5.7.22
> +      container_name: 1-mysql
> +      environment:
> +                - MYSQL_DATABASE=myDB
> +                - MYSQL_USER=fancooler
> +                - MYSQL_PASSWORD=1234
> +                - MYSQL_ROOT_PASSWORD=1234root
> +      volumes:
> +            - ./dump:/docker-entrypoint-initdb.d
> +            - ./db:/var/lib/mysql
> +      networks:
> +            - mynet
> +      ports:
> +            - "3306:3306"
> +    phpmyadmin:
> +     image: phpmyadmin/phpmyadmin
>
> подравни си остъпа да в като в db. Държа много на правилното форматиране :)
> ------------------------------
>
> In dump/01-products.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205530937>:
>
> > @@ -0,0 +1,29 @@
> +DROP DATABASE IF EXISTS mytask;
> +
> +CREATE DATABASE mytask;
> +
> +USE mytask;
> +
> +CREATE TABLE IF NOT EXISTS `products` (
> +	`product_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`title` VARCHAR(30)  NOT NULL DEFAULT '',
> +	`price` DECIMAL(7,2)  NOT NULL DEFAULT '99999.99',
>
> чудесно. Пари трябва да се представят с decimal, а не с float
> ------------------------------
>
> In dump/01-products.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205531057>:
>
> > @@ -0,0 +1,29 @@
> +DROP DATABASE IF EXISTS mytask;
> +
> +CREATE DATABASE mytask;
> +
> +USE mytask;
> +
> +CREATE TABLE IF NOT EXISTS `products` (
> +	`product_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`title` VARCHAR(30)  NOT NULL DEFAULT '',
> +	`price` DECIMAL(7,2)  NOT NULL DEFAULT '99999.99',
> +	`description` MEDIUMTEXT NOT NULL,
> +	`date_of_creation` DATETIME NOT NULL,
> +	PRIMARY KEY (`product_id`))
> +	ENGINE = InnoDB CHARSET=utf8 COLLATE utf8_swedish_ci;
>
> виж дали можеш да използваш utf8mb4
> ------------------------------
>
> In dump/01-products.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205531429>:
>
> > +CREATE DATABASE mytask;
> +
> +USE mytask;
> +
> +CREATE TABLE IF NOT EXISTS `products` (
> +	`product_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`title` VARCHAR(30)  NOT NULL DEFAULT '',
> +	`price` DECIMAL(7,2)  NOT NULL DEFAULT '99999.99',
> +	`description` MEDIUMTEXT NOT NULL,
> +	`date_of_creation` DATETIME NOT NULL,
> +	PRIMARY KEY (`product_id`))
> +	ENGINE = InnoDB CHARSET=utf8 COLLATE utf8_swedish_ci;
> +				
> +INSERT INTO `products` (`title`, `price`, `description`, `date_of_creation`)
> +VALUES	('Shoes', 119.99, 'An item of footwear intended to protect and comfort the human foot.', NOW()),
> +		('Blouse', 42.75, 'A loose-fitting upper garment that was formerly worn by women and children.', NOW()),
>
> вместо now() вкарай някакви различни дати. Може и да не са ръчни, просто
> NOW() - INTERVAL 1 DAY, NOW - INTERVAL 2 DAY и т.н.
> ------------------------------
>
> In dump/01-products.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205531777>:
>
> > @@ -0,0 +1,29 @@
> +DROP DATABASE IF EXISTS mytask;
> +
> +CREATE DATABASE mytask;
> +
> +USE mytask;
> +
> +CREATE TABLE IF NOT EXISTS `products` (
> +	`product_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
>
> Много добре! Челочисленият тип за сурогатния ключ е добре да е
> неотрицателно число, понеже MySQL никога няма да генерира отрицателни числа.
> ------------------------------
>
> In dump/02-users.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205531955>:
>
> > @@ -0,0 +1,28 @@
> +CREATE TABLE IF NOT EXISTS `users` (
> +	`user_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`first_name` VARCHAR(20)  NOT NULL DEFAULT '',
> +	`last_name` VARCHAR(30)  NOT NULL DEFAULT '',
> +	`address` VARCHAR(255)  NOT NULL DEFAULT '',
> +	`country_code` varchar(2) NOT NULL DEFAULT '',
>
> Този стринг винаги ще съдържа два символа. Виж дали има по-подходящ тип.
> ------------------------------
>
> In dump/02-users.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205532503>:
>
> > @@ -0,0 +1,28 @@
> +CREATE TABLE IF NOT EXISTS `users` (
> +	`user_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`first_name` VARCHAR(20)  NOT NULL DEFAULT '',
> +	`last_name` VARCHAR(30)  NOT NULL DEFAULT '',
> +	`address` VARCHAR(255)  NOT NULL DEFAULT '',
>
> В заданието "adress" е уловка. Нищо. Моля те добави отделно:
>
>    - ПК - няма как да е число. Английските ПК са стрингове. Приеми, че
>    дължината на английския ПК е най-дългия възможен ПК.
>    - град
>
> ------------------------------
>
> In dump/03-orders.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205532962>:
>
> > @@ -0,0 +1,37 @@
> +CREATE TABLE IF NOT EXISTS `orders` (
> +	`order_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`order_date` DATETIME NOT NULL,
> +	`user_id` INT UNSIGNED NOT NULL,
> +	`amount` DECIMAL(7,2) NOT NULL DEFAULT '99999.99',
>
> Цените са DECIMAL(7,2). Няма как сумата да е само 7,2 . Разшири малко типа.
> ------------------------------
>
> In dump/03-orders.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205533239>:
>
> > +	PRIMARY KEY(`order_id`),
> +	CONSTRAINT `FK_userorder` FOREIGN KEY (`user_id`) REFERENCES users (`user_id`))
> +	ENGINE = InnoDB CHARSET=utf8 COLLATE utf8_swedish_ci;
> +	
> +INSERT INTO `orders` (`order_date`, `user_id`, `amount`)
> +VALUES				 ('2018-07-19 10:21:34', 1, 51.56),
> +					 ('2018-07-19 12:42:13', 2, 269.70),
> +					 ('2018-07-19 13:57:20', 3, 85.50),
> +					 ('2018-07-20 11:20:07', 4, 479.96),
> +					 ('2018-07-20 14:44:30', 5, 1399.95),
> +					 ('2018-07-20 16:25:09', 6, 88.50),
> +					 ('2018-07-20 17:47:40', 7, 258.58),
> +					 ('2018-07-20 18:00:12', 8, 12.89),
> +					 ('2018-07-21 08:12:43', 9, 30.98),
> +					 ('2018-07-21 10:36:28', 10, 206.67),
> +					 ('2018-07-21 13:55:21', 11, 199.98),
>
> разпръсни малко повече датите на поръчките, за да може заявката за
> последния месец да филтрира част от резултатите.
> ------------------------------
>
> In dump/04-orderLines.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205533416>:
>
> > @@ -0,0 +1,37 @@
> +CREATE TABLE IF NOT EXISTS `orderlines` (
> +	`orderlines_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`order_id` INT UNSIGNED NOT NULL,
> +	`price_each` DECIMAL(7,2) NOT NULL DEFAULT '99999.99',
>
> Бих го нарекал item_price
> ------------------------------
>
> In dump/04-orderLines.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205533983>:
>
> > @@ -0,0 +1,37 @@
> +CREATE TABLE IF NOT EXISTS `orderlines` (
> +	`orderlines_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`order_id` INT UNSIGNED NOT NULL,
> +	`price_each` DECIMAL(7,2) NOT NULL DEFAULT '99999.99',
>
> Един допълнителен индекс order_id, product_id ще помогне на определени
> заявки
> ------------------------------
>
> In dump/04-orderLines.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205534100>:
>
> > @@ -0,0 +1,37 @@
> +CREATE TABLE IF NOT EXISTS `orderlines` (
> +	`orderlines_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
> +	`order_id` INT UNSIGNED NOT NULL,
> +	`price_each` DECIMAL(7,2) NOT NULL DEFAULT '99999.99',
> +	`product_id` INT UNSIGNED NOT NULL,
> +	`quantity` INT UNSIGNED NOT NULL,
>
> добре
> ------------------------------
>
> In dump/05-printing_last_hour.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205534367>:
>
> > @@ -0,0 +1,3 @@
> +/* Printing the list of products, created during the last hour. */
> +																						
> +SELECT title FROM products WHERE date_of_creation > SUBDATE( NOW(), INTERVAL 1 HOUR);
>
> правилно. Може и DATE_SUB() или направо с израз : NOW - INTERVAL 1 HOUR
> ------------------------------
>
> In dump/06-printing_code.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205534463>:
>
> > @@ -0,0 +1,3 @@
> +/* Printing the list of users sorted by country code.*/
> +																						
> +SELECT first_name, last_name, country_code FROM users ORDER BY country_code;
>
> правилно
> ------------------------------
>
> In dump/07-printing_orders.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205534773>:
>
> > @@ -0,0 +1,3 @@
> +/* Printing a simple list of orders.*/
> +
> +SELECT order_id, user_id, order_date, product_id, price_each, amount FROM orderlines INNER JOIN orders using (order_id);
>
> в тази заявка не ти трябва orderlines. Простият списък идва само от
> orders. Заданието е :
> "Printing a simple list of orders (no order lines)."
> ------------------------------
>
> In dump/04-orderLines.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205536144>:
>
> > @@ -0,0 +1,37 @@
> +CREATE TABLE IF NOT EXISTS `orderlines` (
> +	`orderlines_id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
>
> хубаво е да добавиш и позиция на линията в поръчката.
> ------------------------------
>
> In dump/08-printing_orderlines.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205536337>:
>
> > @@ -0,0 +1,3 @@
> +/* 	Printing the list orderlines with combined order information and user information.*/
> +
> +SELECT order_id, price_each, product_id, quantity, order_date, user_id, amount, first_name, last_name, address, country_code FROM orderlines INNER JOIN orders using (order_id) INNER JOIN users using (user_id) ORDER BY order_id;
>
> В SQL, ако не е зададена сортировка, не може да се гарантира че всеки път
> резултатът ще е същият. В случая, се изписват orderlines и се сортира по
> order_id. Като добавиш orderlines.position, я използвай също за сортирането.
> ------------------------------
>
> In dump/09-printing_number_of_orders.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205537383>:
>
> > @@ -0,0 +1,3 @@
> +/* 	Printing the list of users with number of orders they have placed.*/
> +
> +SELECT first_name, last_name, users.user_id, COUNT(*) AS number_of_orders FROM users, orders WHERE users.user_id = orders.user_id GROUP BY first_name, last_name, user_id ORDER BY users.user_id;
>
> Бих сложил users.user_id отпред, вместо след имената.
> Възможно е и използването на INNER JOIN с USING или ON клауза. По-добре
> показва как се навързват таблиците, а WHERE клаузата остава за филтрирането
> (selection).
> ------------------------------
>
> In dump/10-printing_spent_in_total.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205537775>:
>
> > @@ -0,0 +1,3 @@
> +/* Printing the list of users with a column of how much money they have spent in this shop. */
> +
> +SELECT SUM(orders.amount) total, users.first_name, users.last_name FROM orders, users WHERE orders.order_id = users.user_id GROUP BY first_name, last_name ORDER BY total;
>
> Групирай по друга колона. В магазина може да има два пъти Борис Христов.
> ------------------------------
>
> In dump/09-printing_number_of_orders.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205538509>:
>
> > @@ -0,0 +1,3 @@
> +/* 	Printing the list of users with number of orders they have placed.*/
> +
> +SELECT first_name, last_name, users.user_id, COUNT(*) AS number_of_orders FROM users, orders WHERE users.user_id = orders.user_id GROUP BY first_name, last_name, user_id ORDER BY users.user_id;
>
> Не трябва да групираш по име и фамилия
> ------------------------------
>
> In dump/11-printing_more_than_three_users.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205539067>:
>
> > @@ -0,0 +1,3 @@
> +/* Printing the list of countries that have more than 3 users, sorted by this number in descending order. */
> +																						
> +SELECT country_code, COUNT(*) AS mttu FROM users GROUP BY country_code HAVING mttu >= 3 ORDER BY mttu DESC;
>
> more than three users = mttu бих го кръстил total_users
> ------------------------------
>
> In dump/12-printing_orders_per_country.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205539500>:
>
> > @@ -0,0 +1,3 @@
> +/* Printing the list of countries with the number of orders per country, sorted by this number in descending order. */
> +																						
> +SELECT country_code, COUNT(*) AS nopc FROM users, orders WHERE users.user_id = orders.user_id GROUP BY country_code ORDER BY nopc DESC;
>
> Тук пак може да се използва INNER JOIN. Може, не е задължително.
> ------------------------------
>
> In dump/15-printing_most_ordered.sql
> <https://github.com/rastepanyan/training/pull/2#discussion_r205539689>:
>
> > @@ -0,0 +1,2 @@
> +/* Printing the 10 most ordered products of all times. */
> +SELECT product_id, SUM(quantity) AS most_ordered FROM orderlines GROUP BY product_id ORDER BY most_ordered DESC LIMIT 10;
>
> правилно
> ------------------------------
>
> In www/index.php
> <https://github.com/rastepanyan/training/pull/2#discussion_r205542652>:
>
> > @@ -0,0 +1,34 @@
> +<html>
>
> още не сме стигнали до PHP, но добре :)
> За по-голямо удобство може да направиш следното. Започни файла с PHP код,
> където правиш връзката, пускаш заявката, проверяваш за грешки и чак тогава
> затваряш PHP кода и започва HTML-a. В самия HTML пак отваряш PHP за да
> можеш да врътнеш цикъла за резултатите. Така ще имаш минимално количество
> PHP код в HTML-a. Реално погледнато тогава HTML-a може да се изнесе в
> отделен файл, т.е. да е шаблон/темплейт, несъдържащ бизнес логика.
> Така кодът ще е на ниво PHP до 2005-та година, преди появяването на
> фреймуърци, но пак виждаш основните компоненти и разделянето на бизнес
> логика от презентация.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rastepanyan/training/pull/2#pullrequestreview-140819581>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVsTfHo5ZJyB_wtK4Vcc1JWce0mOysGeks5uKf6ngaJpZM4ViKA->
> .
>

rastepanyan committed:
Update 2

01-PRODUCTS.SQL

1. Fix: Generate different Date/Time using ‘NOW() - INTERVAL * DAY/MONTH/HOUR/MINUTE’.


04-ORDERLINES.SQL

1. Correct some ‘product_id’ values.


07-PRINTING_ORDERS.SQL

1. Fix: Now prints order list correctly.


08-PRINTING_ORDERLINES.SQL

1. Print ‘post_code’ and ‘city’ as a result too.




Add:

13-PRINTING_TOP_3_BUYERS.SQL
14-PRINTING_TURNOVER.SQL
16-PRINTING_NEVER_ORDERED.SQL


Note: 

Need to put more and different month names in 03-ORDERS.SQL to get proper result from 14-PRINTING_TURNOVER.SQL

rastepanyan committed:
Re-Format docker-compose.yml again

rastepanyan committed:
Merge branch 'mysql-task' of https://github.com/rastepanyan/training into mysql-task

rastepanyan committed:
Add re-formatted docker-compose.yml

rastepanyan committed:
Delete unnecessary directory

andrey-dnh commented on dump/09-printing_number_of_orders.sql:
```diff
@@ -0,0 +1,3 @@
+/* 	Printing the list of users with number of orders they have placed.*/
+
+SELECT users.user_id, first_name, last_name, COUNT(*) AS number_of_orders FROM users INNER JOIN orders WHERE users.user_id = orders.user_id GROUP BY user_id ORDER BY users.user_id;
```
Това, което имах предвид е, че : 
```
SELECT users.user_id, first_name, last_name, COUNT(*) AS number_of_orders
FROM users INNER JOIN orders USING(user_id)
GROUP BY user_id
ORDER BY users.user_id
```
или
```
SELECT users.user_id, first_name, last_name, COUNT(*) AS number_of_orders
FROM users INNER JOIN orders ON users.user_id = orders.user_id
GROUP BY user_id
ORDER BY users.user_id
```
Условието от WHERE клаузата отива в JOIN-a. Kato ON се използва за по-сложни неща и когато имената на колоните не съвпадат, а USING е за навързване само по една колона.
Така става много по-ясно кои условия са за навързване на таблиците и кои за филтриране.

andrey-dnh commented on dump/13-printing_top_3_buyers.sql:
```diff
@@ -0,0 +1,8 @@
+/* Printing the top 3 buyers (id, names, country) in the shop based on turnover */
+
+SELECT users.user_id, CONCAT(first_name, " ", last_name) AS names, country_code, SUM(item_price * quantity) AS turnover
+FROM users INNER JOIN orders USING (user_id) INNER JOIN orderlines USING (order_id)
```
ето, тук си го направил така :)

andrey-dnh commented on dump/13-printing_top_3_buyers.sql:
```diff
@@ -0,0 +1,8 @@
+FROM users INNER JOIN orders USING (user_id) INNER JOIN orderlines USING (order_id)
+GROUP BY users.user_id
+ORDER BY turnover 
+DESC 
```
DESC е за ORDER BY и би трябвало да стои на неговия ред

andrey-dnh commented on dump/16-printing_never_ordered.sql:
```diff
@@ -0,0 +1,8 @@
+/* Printing the products that have never been ordered */
+
+SELECT title, 
```
Виж как може да се реши без подзaявка.
https://www.codeproject.com/KB/database/Visual_SQL_Joins/Visual_SQL_JOINS_V2.png 
това може да ти е от помощ.

andrey-dnh replied:
```
Server version: 5.7.22 MySQL Community Server (GPL)

Copyright (c) 2000, 2018, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> select now() - interval 1 day;
+------------------------+
| now() - interval 1 day |
+------------------------+
| 2018-07-30 18:39:32    |
+------------------------+
1 row in set (0.00 sec)
```
Странно

andrey-dnh commented on dump/14-printing_turnover.sql:
```diff
@@ -0,0 +1,8 @@
+FROM orderlines 
+INNER JOIN orders using (order_id)
+WHERE order_date >= DATE_SUB(NOW(), INTERVAL 10 MONTH)
+GROUP BY month
```
нека се принтира и годината, освен месеца и сортирането да бъде по тях DESC, DESC.

rastepanyan committed:
Add fixed index.php and re-formatted docker-compose.yml

rastepanyan committed:
Same as index.php

andrey-dnh commented on www/index.php:
```diff
@@ -1,34 +1,37 @@
+   $password="1234root";
+   $db="myDB";
+   	
+   $conn=m ysqli_connect($servername, $username, $password, $db);
```
според PSR-1 / 2 трябва да имаш интервали около оператора за присвояване.
Отделно имаш интервал в името на функцията

andrey-dnh commented on www/index.php:
```diff
@@ -1,34 +1,37 @@
+   if (!$conn) {
+   	die( "Failed! " . mysqli_connect_error());
+   } 
+   echo "Connection established";
```
Това предполагам е временно дебъг съобщение

rastepanyan committed:
New fixes

rastepanyan committed:
New fixes2

rastepanyan committed:
New fixes2a
